### PR TITLE
Switch to c++17 and leverage shared_mutex in tesseract_environment

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

File: tesseract/tesseract_environment/src/core/environment.cpp [hunk 1]
```diff
@@ -443,23 +443,24 @@ void Environment::removeAllowedCollision(const std::string& link_name)
...
   tesseract_scene_graph::AllowedCollisionMatrix::ConstPtr acm = scene_graph_->getAllowedCollisionMatrix();
 
   return acm;
 }
 
 tesseract_scene_graph::Joint::ConstPtr Environment::getJoint(const std::string& name) const
 {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::shared_lock<std::shared_mutex> lock(mutex_);
+  ;
```
Comment (mpowelson) [inline tesseract/tesseract_environment/src/core/environment.cpp:455]: extra `;`

**Review by mpowelson — APPROVED** [review]
One inconsequential change

**Comment by mpowelson** [comment]
> One inconsequential change

Actually, just looked at CI. That's not good.

**Comment by mpowelson** [comment]
Yeah we may be more careful than I was thinking because have cases all over the place where unique locked things are calling unique locked functions.

**Comment by Levi-Armstrong** [comment]
Yep, I need to look into this further to see where the double lock is occurring.

**Comment by Levi-Armstrong** [comment]
This is included in #544

**Pull request closed without merging**